### PR TITLE
Support Azure Pipelines

### PR DESCRIPTION
**Pull request by martincostello**
This PR adds support for public Azure Pipelines builds.

It also fixes a typo in an enum value's name I found while I was implementing it.

Some open questions from while I was implementing this:
  1. Should an extra query string parameter be added to filter by build definition? Unlike the other CI services, Azure Pipelines supports having lots of different build definitions that could do completely different things, meaning that what's being shown in the graph might not be like-for-like. The API has a `definitions` parameter that could be used to feed in the IDs from a query string to filter the chart to just a specific build definition(s).
  1. Should a maximum number be specified? I noticed Circle CI sets a maximum value.
  1. I haven't tested the behaviour for a non-pubic project or whether or passing through the new API token support and whether that would work for non-public projects. Does the code I've based this on already handle 401/403 nicely, or can that be added at a later date if-and-when needed?
  1. Is `azurepipelines` OK for the name?

Below is a (zoomed-out) screenshot of the test page running locally on my laptop.

image

### CODE REVIEW

**Comment by dustinmoris** [comment]
Wow amazing! I was wanting to add support for Azure Pipelines myself over the next few days but you have just made my life so much easier!!!

> Should an extra query string parameter be added to filter by build definition? Unlike the other CI services, Azure Pipelines supports having lots of different build definitions that could do completely different things, meaning that what's being shown in the graph might not be like-for-like. The API has a definitions parameter that could be used to feed in the IDs from a query string to filter the chart to just a specific build definition(s).

Yes, that is fine. If it makes sense for Azure Pipelines then we should have that additional query parameter for the Azure Pipelines API endpoint.

> Should a maximum number be specified? I noticed Circle CI sets a maximum value.

This might just be CircleCI specific. All CI systems are slightly different in what capabilities their APIs have. The most preferred option is if the API already exposes a way of filtering builds based on the filters which CI-BuildStats offer (branch, pull request, number of builds), because then we can just make a single request to their API knowing that the data we will get back already matches all filters. However, if the API doesn't for example offer a filter for pull requests, then I normally query a bit more with the assumption that after filtering out the PRs there will hopefully still be enough builds left which match the users desired `buildCount`, otherwise I have to fire off a second request. I think this is what I do for CircleCI and I set probably a hard limit to avoid too many requests.

> I haven't tested the behaviour for a non-pubic project or whether or passing through the new API token support and whether that would work for non-public projects. Does the code I've based this on already handle 401/403 nicely, or can that be added at a later date if-and-when needed?

This can be added afterwards. I haven't publicised the support for private repos yet, because I want to make it a bit fancier before making it completely public. Currently I consider it an "early adopter test" feature which probably will become better over the next few weeks followed by a proper blog post explaining how it works.

> Is azurepipelines OK for the name?

Yes. It's a bit wordy, but that's not our fault ^^

**Comment by dustinmoris** [comment]
Do you know what, I wasn't actually able to find any official documentation around their API when googling for it. Have you got a link where I could see what the API supports at the moment? Many thanks!

**Comment by martincostello** [comment]
Sure, here’s what I was working from: https://docs.microsoft.com/en-us/rest/api/vsts/build/builds/list

**Comment by martincostello** [comment]
I've updated the code to filter by `definitionId={int}` as an optional query string parameter and added an example for it to the tests.

I had a quick look for PR filtering, but it looks like you can only filter to whether it _is_ one with `reasonFilter=pullRequest`, but I can't find away to negate that.

From a quick test (I used `https://dev.azure.com/dnceng/public/_apis/build/builds?branchName=&api-version=4.1`, which is the Pipelines for the .NET Core OSS stuff) it looks like the default value of `$top` if it isn't specified is 1,000. We could probably set a default maximum a bit like for Circle CI I guess to return something more reasonable before the PRs are filtered out by this app. Say 100?

**Comment by martincostello** [comment]
By the way, I'm not sure how you run this to do dev, but I've just been doing `dotnet run` + Chrome. However, when I start the app it crashes because of something to do with the CloudFlare IP whitelist having an invalid CIDR and also I'm not sure what the API token should be to view the tests page.

To work around that I've been just commenting out the relevant bits (and remembering not to commit them) to test locally. Is there a bug, or some instructions for what I should be doing to not need to do that?

**Comment by martincostello** [comment]
Oh yeah, and I’m a F# n00b, so if I’ve done anything clunky, sorry!

**Comment by dustinmoris** [comment]
Maybe something like 4 * the selected buildCount but no more than 200 total or something like that?

**Comment by dustinmoris** [comment]
Hi, it all looks great! I am currently working on the final touches (docs, etc.) for a new Giraffe release which will have a few substansial perf improvements for apps which use the GiraffeViewEngine and JSON endpoints and I would like to get that finished today/tomorrow first and then apply that change together with this PR to CI-BuildStats so I have to roll out only one new version this week.

I anticipate to have this feature go live on Wednesday hopefully. I thought I'd give you a quick update in case you wondered why I haven't merged and pushed live yet :)

**Comment by martincostello** [comment]
Awesome, no worries!

**Comment by dustinmoris** [comment]
Hi,

So by looking at Azure Pipelines it really seems like that each project can have different build definitions (= definitionId), right?

A graph which isn't limited to a specific definition id doesn't really make any sense, as each definition can do something completely different, so this should be almost mandatory like the account and project name, correct?

Do you know how the concept of different definition ids plays together with a repository which has a build definiton in a yaml file? Would a project with multiple build definitions have multiple yaml files?

**Comment by martincostello** [comment]
So as I understand it, each definition _may_ have a YAML file, but you could _also_ configure multiple definitions off a single YAML file (e.g. if you put conditional logic in it).

**Comment by dustinmoris** [comment]
Right... ok... thanks!

**Comment by dustinmoris** [comment]
Hi, I just committed a minor change where I'd propose to make the definition ID a mandatory route argument for the azure pipelines route only. I think a graph which renders a mix of random builds doesn't make much sense otherwise... I don't know, are you ok with that chagne or do you see some value in having it as an optional query parameter? (I made the minor change to see if it would work in principal)

**Comment by martincostello** [comment]
I think that’s reasonable.

The argument for not having it mandatory is that for a simple hobby project it’s one less thing to configure and it’s consistent with the other three providers, but in the case of a non-trivial one you just end up with a mess that makes no sense until you filter it.

**Comment by martincostello** [comment]
Thanks for merging this Dustin! 😃

**Comment by dustinmoris** [comment]
Thanks for implementing it ;)
